kernel: fix errno access for user mode

tyrantlizard77 opened the pull request:
The errno "variable" is required to be thread-specific.
It gets defined to a macro which dereferences a pointer
returned by a kernel function.

In user mode, we cannot simply read/write the thread struct.
We do not have thread-local storage mechanism, so for now
use the lowest address of the thread stack to store this
value, since this is guaranteed to be read/writable by
a user thread.

The downside of this approach is potential stack corruption
if the stack pointer goes down this far but does not exceed
the location, since a fault won't be generated in this case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: fix errno access for user mode

The errno "variable" is required to be thread-specific.
It gets defined to a macro which dereferences a pointer
returned by a kernel function.

In user mode, we cannot simply read/write the thread struct.
We do not have thread-local storage mechanism, so for now
use the lowest address of the thread stack to store this
value, since this is guaranteed to be read/writable by
a user thread.

The downside of this approach is potential stack corruption
if the stack pointer goes down this far but does not exceed
the location, since a fault won't be generated in this case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross reviewed (approved):
Looks clean.  It might be worth futureproofing this by defining the location in a "TLS struct" that you put at the bottom of the stack, and then defining an API to get access to that.

On SMP-enabled platforms, we'd then be able to insert a proper TLS access mechanism (not that this has anything to do with SMP, but it shares an implementation with the SMP CPU struct pointer and is trivial there), falling back to the syscall on other platforms.

lpereira reviewed (approved):
LGTM with a question: is the initialization of `errno_location` guaranteeing the alignment for an `int`?

andrewboie commented:
> LGTM with a question: is the initialization of errno_location guaranteeing the alignment for an int?

Yes, its the lowest memory address of the stack buffer which is at least aligned to 4 bytes (and typically some much larger value)

lpereira commented:
Another question: will this work with stacks that grow up and down?

Commit:
syscalls: pull in arch/cpu.h

Need the implementation of _arch_syscall_invokeX() functions
to avoid build problems in some cases.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: arch.h: move extern "C" after includes

Shouldn't declare this and then pull in headers, fixes
errors like "util.h:41:1: error: template with C linkage"

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arc: arch.h: move extern "C" after includes

Shouldn't declare this and then pull in headers, fixes
error like "util.h:41:1: error: template with C linkage"

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
